### PR TITLE
Protos with no GOPATH

### PR DESCRIPTION
**Pull request by EItanya**
BOT NOTES: 
resolves https://github.com/solo-io/solo-kit/issues/326

### CODE REVIEW

File: api/v1/generate.sh [hunk 1]
```diff
@@ -2,28 +2,45 @@
...
+
+
+ROOT=$(dirname "${BASH_SOURCE[0]}")/../../..
+SOLO_KIT=${ROOT}/solo-kit
 IN=${SOLO_KIT}/api/v1/
 EXTERNAL=${SOLO_KIT}/api/external/
-OUT=${SOLO_KIT}/pkg/api/external/istio/encryption/v1/
+
+# code-generator does work with go.mod but makes assumptions about
+# the project living in $GOPATH/src. To work around this and support
```
Comment (kdorosh) [inline api/v1/generate.sh]: what kinds of assumptions? how are these assumptions leveraged? at some point we'd like to undo some of these assumptions -- or at least properly document them

File: ci/tools.go [hunk 2]
```diff
@@ -13,9 +13,13 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// This package imports things required by build scripts, to force `go mod` to see them as dependencies
 package tools
 
 import (
+	// I had to remove this as the version of gogo.proto it has is out of date,
+	// and therefore breaking our protoc commands.
```
Comment (kdorosh) [inline ci/tools.go]: what is "this"?

File: cloudbuild.yaml [hunk 3]
```diff
@@ -1,130 +1,111 @@
...
-  - name: 'gcr.io/$PROJECT_ID/dep:0.1.5'
-    args: ['ensure']
-    env:
-      - 'PROJECT_ROOT=github.com/solo-io/solo-kit'
-      - 'GIT_SSH_CONFIG=FALSE'
-    id: 'dep'
-    waitFor: ['download-untar-dep-cache']
+# TODO: during this step, prepare-workspace to set up the project so it can be built and tested
+# Dep modifies workspace so that the code is moved into the PROJECT_ROOT
+# All steps after this must set GOPATH
```
Comment (kdorosh) [inline cloudbuild.yaml]: delete

File: cloudbuild.yaml [hunk 4]
```diff
@@ -1,130 +1,111 @@
...
-  # Dep modifies workspace so that the code is moved into the PROJECT_ROOT
-  # All steps after this must set GOPATH
-  - name: 'gcr.io/$PROJECT_ID/dep:0.1.5'
-    args: ['ensure']
-    env:
-      - 'PROJECT_ROOT=github.com/solo-io/solo-kit'
-      - 'GIT_SSH_CONFIG=FALSE'
-    id: 'dep'
-    waitFor: ['download-untar-dep-cache']
+# TODO: during this step, prepare-workspace to set up the project so it can be built and tested
```
Comment (kdorosh) [inline cloudbuild.yaml]: what exactly is the TODO here?

File: cloudbuild.yaml [hunk 5]
```diff
@@ -1,130 +1,111 @@
...
+  - solo-kit
+  - "--repo-sha"
+  - "$COMMIT_SHA"
+  - "--repo-output-dir"
+  - "."
+  env:
+  - 'GIT_SSH_CONFIG=FALSE'
+  id: 'prepare-workspace'
 
+# donwload massive container in parallel
```
Comment (kdorosh) [inline cloudbuild.yaml]: typo,
```suggestion
# download massive container in parallel
```

File: pkg/code-generator/cmd/main.go [hunk 6]
```diff
@@ -58,34 +62,130 @@ type GenerateOptions struct {
 	SkipGenMocks bool
 	// skip generated tests
 	SkipGeneratedTests bool
+	/*
+		Represents the go package which this package would have been in the GOPATH
+		This allows it to be able to maintain compatility with the old solo-kit
+
+		default: current github.com/solo-io/<current-folder>
+		for example: github.com/solo-io/solo-it
```
Comment (kdorosh) [inline pkg/code-generator/cmd/main.go]: typo,
```suggestion
		for example: github.com/solo-io/solo-kit
```

File: pkg/code-generator/cmd/main.go [hunk 7]
```diff
@@ -58,34 +62,130 @@ type GenerateOptions struct {
...
+		if info.IsDir() {
+			return nil
+		}
+		if !(strings.HasSuffix(pbgoFile, ".pb.go") || strings.HasSuffix(pbgoFile, ".pb.hash.go")) {
+			return nil
+		}
+
+		dest := strings.TrimPrefix(pbgoFile, filepath.Join(descriptorOutDir, r.Opts.PackageName))
+		dest = strings.TrimPrefix(dest, "/")
+		// dest = filepath.Join(relativeRoot, dest)
```
Comment (kdorosh) [inline pkg/code-generator/cmd/main.go]: remove

File: ci/tools.go (hunk 2 above)
Comment (EItanya) [reply ci/tools.go]: removed, no longer valid

**Comment by EItanya** [comment]
docs-gen isn't quite working, updating now

File: .gitignore [hunk 8]
```diff
@@ -19,4 +19,4 @@
 vendor/
 _output/
 /solo-kit-gen
-/doc-gen-test-*
+**/doc-gen-test-*
```
Comment (ilackarms) [inline .gitignore]: what's the `**` syntax  mean?

File: README.md [hunk 9]
```diff
@@ -13,34 +13,12 @@ A collection of code generation and libraries to for API development.
...
-- clone repo to gopath 
-- gather dependencies: `dep ensure -v`
+- clone repo locally
+- gather dependencies: `go mod download`
 - use binary produced with `make solo-kit-gen` or import `cmd.Run(...)` into your own code gen code 
 
 ## Usage
 - re-run whenever you change or add an api (.proto file)
-- api objects generated from messages defined in protobuf files with magic comments prefixed with `@solo-kit`
+- api objects generated from messages defined in protobuf files which have
```
Comment (ilackarms) [inline README.md]: good idea updating this, but comment is unfinished

File: .gitignore (hunk 8 above)
Comment (EItanya) [reply .gitignore]: Any folder from root, because it now lives in a nested folder structure, it needs to be ignored there rather than just at the root

File: README.md (hunk 9 above)
Comment (EItanya) [reply README.md]: I will be doing a full once over of the docs, as well as updating for everything new

File: changelog/v0.12.0/gomod.yaml [hunk 10]
```diff
@@ -0,0 +1,4 @@
+changelog:
+- type: BREAKING_CHANGE
```
Comment (ilackarms) [inline changelog/v0.12.0/gomod.yaml]: why is this a breaking change?

Comment (EItanya) [reply changelog/v0.12.0/gomod.yaml]: We decided it was a breaking change because it completely changes the way it functions

File: api/v1/generate.sh [hunk 11]
```diff
@@ -2,28 +2,46 @@
...
 
-ROOT=${GOPATH}/src
-SOLO_KIT=${ROOT}/github.com/solo-io/solo-kit
+set -o errexit
+set -o nounset
+set -o pipefail
+
+
+# The following script is used to generate the solo-kit protos.
+# This script will work both in and our of the GOPATh, however, it does assume that the imported protos will be
```
Comment (ashleywang1) [inline api/v1/generate.sh]: GOPATh -> GOPATH

File: cloudbuild.yaml [hunk 12]
```diff
@@ -1,130 +1,109 @@
...
+  env:
 
 timeout: 1500s
 tags: ['solo-kit']
 options:
   machineType: 'N1_HIGHCPU_32'
-  env:
-  - "GO111MODULE=on"
\ No newline at end of file
+  volumes:
```
Comment (ashleywang1) [inline cloudbuild.yaml]: The GO111MODULE=auto by default for go 1.13, but the auto setting now activates the module-aware mode of the go command whenever the current working directory contains, or is below a directory containing, a go.mod file — even if the current directory is within GOPATH/src. This change simplifies the migration of existing code within GOPATH/src and the ongoing maintenance of module-aware packages alongside non-module-aware importers.

yay

File: pkg/code-generator/docs_gen_test.go [hunk 13]
```diff
@@ -73,13 +86,36 @@ var _ = Describe("DocsGen", func() {
 				ApiDir:  hugoApiDir,
 			},
 		}
+
 		// Run code gen
 		opts := cmd.GenerateOptions{
-			CustomImports: []string{filepath.Join(tempDir, "..", "vendor/github.com/solo-io/protoc-gen-ext")},
-			RelativeRoot:  tempDir,
+			// CustomImports: []string{
```
Comment (ashleywang1) [inline pkg/code-generator/docs_gen_test.go]: delete this?

File: pkg/protodep/protodep.proto [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+message Local {
+    repeated string patterns = 1;
+}
+
+message GoModImport {
+    repeated string patterns = 1 [(validate.rules).repeated = { min_items: 1}];
+    string package = 2 [(validate.rules).string = { min_len: 1}];
+}
+
+message GitImport {
```
Comment (rickducott) [inline pkg/protodep/protodep.proto]: nit: should not need to merge in more api than we have implemented

File: pkg/protodep/gomod.go [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	modPackageReader, err := modutils.GetCurrentPackageList()
+	if err != nil {
+		return nil, err
+	}
+
+	// split list of pacakges from cmd by line
```
Comment (mitchdraft) [inline pkg/protodep/gomod.go]:
```suggestion
	// split list of packages from cmd by line
```

File: pkg/protodep/gomod.go [hunk 16]
```diff
@@ -0,0 +1,331 @@
+package protodep
+
```
Comment (rickducott) [inline pkg/protodep/gomod.go]: Would have expected this library in a standalone PR, this PR is already quite large yet the testing for the library seems too minimal if we’re to use this everywhere

File: pkg/protodep/gomod.go [hunk 17]
```diff
@@ -0,0 +1,331 @@
...
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// gather up all packages for a given go module
+// currently this function uses the cmd `go list -m all` to figure out the list of dep
+func (m *goModFactory) gather(opts goModOptions) ([]*Module, error) {
+	matchOptions := opts.MatchOptions
```
Comment (rickducott) [inline pkg/protodep/gomod.go]: nit: this function is too long ( > 100 lines)

File: pkg/protodep/gomod.go [hunk 18]
```diff
@@ -0,0 +1,331 @@
...
+	DefaultMatchPatterns = []string{ProtoMatchPattern, SoloKitMatchPattern}
+
+	// matches ext.proto for solo hash gen
+	ExtProtoMatcher = &GoModImport{
+		Package:  "github.com/solo-io/protoc-gen-ext",
+		Patterns: []string{"extproto/*.proto"},
+	}
+
+	// matches validate.proto which is needed by envoy protos
+	ValidateProtoMatcher = &GoModImport{
```
Comment (mitchdraft) [inline pkg/protodep/gomod.go]: nit: might rename this to indicate relation to envoy since we're exporting it

File: pkg/protodep/gomod.go [hunk 19]
```diff
@@ -0,0 +1,331 @@
...
+
+	err = m.copy(mods)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// gather up all packages for a given go module
+// currently this function uses the cmd `go list -m all` to figure out the list of dep
```
Comment (mitchdraft) [inline pkg/protodep/gomod.go]: I appreciate this note but it's not clear where `go list -m all` is called - might be worth indicating where this is happening

File: pkg/utils/modutils/mod.go [hunk 20]
```diff
@@ -0,0 +1,75 @@
+package modutils
+
+import (
+	"bufio"
+	"bytes"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/rotisserie/eris"
```
Comment (mitchdraft) [inline pkg/utils/modutils/mod.go]: nit: I feel like we have a lot of libs that do this or similar and am wary of adding another. I think this was to avoid importing from go-utils. If so, might be worth extracting our preferred error lib to its own repo.

File: pkg/code-generator/docs_gen_test.go [hunk 21]
```diff
@@ -61,7 +67,14 @@ var _ = Describe("DocsGen", func() {
...
 		buf = &bytes.Buffer{}
-		err = projectConfigFile().Execute(buf, projectConfigDocsDir)
+		err = projectConfigFile().Execute(buf, struct {
+			Dir       string
+			GoPackage string
+		}{
+			Dir:       projectConfigDocsDir,
+			GoPackage: relativePathToTempDir,
+		})
+		// err = projectConfigFile().Execute(buf, projectConfigDocsDir)
```
Comment (mitchdraft) [inline pkg/code-generator/docs_gen_test.go]: stray comment

File: pkg/protodep/README.md [hunk 22]
```diff
@@ -0,0 +1,53 @@
+# protodep
```
Comment (mitchdraft) [inline pkg/protodep/README.md]: nit: might be worth extracting the todos/forward-looking comments to a separate section at the end. When reading this I'd like to know how to use the tool as it stands now

File: pkg/utils/modutils/mod.go (hunk 20 above)
Comment (EItanya) [reply pkg/utils/modutils/mod.go]: It's not, and actually I have good reason. I looked into this lib, and it really delivers on the promise of errors in go 1.13. I'm open to removing it, but tbh I'd rather swap this out with my implementation from go-utils

File: api/v1/generate.sh [hunk 23]
```diff
@@ -2,28 +2,46 @@
...
+# This script will work both in and our of the GOPATH, however, it does assume that the imported protos will be
+# available in the root level vendor folder. This script will be run as part of `make generated-code` so there
+# should be no need to run it otherwise. `make generated-code` will also vendor the necessary protos.
+ROOT=$(dirname "${BASH_SOURCE[0]}")/../../..
+SOLO_KIT=${ROOT}/solo-kit
 IN=${SOLO_KIT}/api/v1/
 EXTERNAL=${SOLO_KIT}/api/external/
-OUT=${SOLO_KIT}/pkg/api/external/istio/encryption/v1/
+
+TEMP_DIR=$(mktemp -d)
```
Comment (ilackarms) [inline api/v1/generate.sh]: mktemp might have different parameters on linux, worth checking

File: changelog/v0.12.0/gomod.yaml (hunk 10 above)
Comment (ilackarms) [reply changelog/v0.12.0/gomod.yaml]: the description for the PR is confusing/incomplete. can you give a more detailed account?

File: ci/tools.go [hunk 24]
```diff
@@ -13,9 +13,11 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// This package imports things required by build scripts, to force `go mod` to see them as dependencies
 package tools
```
Comment (ilackarms) [inline ci/tools.go]: please add a comment to explain the purpose of this file and what the underscore imports do

File: generate.go [hunk 25]
```diff
@@ -16,6 +17,22 @@ func main() {
...
+					ImportType: &protodep.Import_GoMod{GoMod: protodep.ExtProtoMatcher},
+				},
+				{
+					ImportType: &protodep.Import_GoMod{GoMod: protodep.ValidateProtoMatcher},
+				},
+				{
+					ImportType: &protodep.Import_GoMod{GoMod: protodep.GogoProtoMatcher},
+				},
+			},
+		},
```
Comment (ilackarms) [inline generate.go]: i am concerned with the complexity of this new config for solo-kit generate. unlike the existing options which are pretty simple / easy to understand for a basic solo-kit user, the options provided here require an in-depth understanding of the new `protodep` pacakge.

I think for the usability of Solo-Kit we need to simplify this process for our solo-kit projects, so we do not wind up with magical/overcomplex `generate.go` in all of our projects

File: pkg/code-generator/codegen/generator.go [hunk 26]
```diff
@@ -188,6 +191,19 @@ func generateKubeFilesForProject(project *model.Project) (code_generator.Files,
...
+	} {
+		content, err := generateProjectFile(project, tmpl)
+		if err != nil {
+			return nil, errors.Wrapf(err, "internal error: processing template '%v' for project %v failed", tmpl.ParseName, project.ProjectConfig.Name)
+		}
+		v = append(v, code_generator.File{
+			Filename: filepath.Join("kube", suffix),
+			Content:  content,
+		})
+	}
```
Comment (ilackarms) [inline pkg/code-generator/codegen/generator.go]: these are a LOT of changes to squeeze into a single PR. i understand the reasoning for wanting to do it together, but it makes reviewing and vetting each change significantly harder. Is there a good reason not to split them into a separate PR?
 
what is the reasoning behind generating these new files?

File: pkg/code-generator/codegen/templates/kube/doc_template.go [hunk 27]
```diff
@@ -11,7 +11,7 @@ var DocTemplate = template.Must(template.New("kube_doc").Funcs(templates.Funcs).
 
 /* go:generate command for Kubernetes code-generator currently disabled, run the following manually (or uncomment and remove the minus):
 	
-- //go:generate $GOPATH/src/k8s.io/code-generator/generate-groups.sh all "{{ .ProjectConfig.GoPackage }}/kube/client" "{{ .ProjectConfig.GoPackage }}/kube/apis" {{ .ProjectConfig.Name }}:{{ .ProjectConfig.Version }}
+- //go:generate bash ../../../hack/update-codegen.sh
```
Comment (ilackarms) [inline pkg/code-generator/codegen/templates/kube/doc_template.go]: this is not a safe assumption... we do not know where this file will live relative to `hack` directory

File: pkg/protodep/protodep.proto [hunk 28]
```diff
@@ -0,0 +1,27 @@
+syntax = "proto3";
```
Comment (joekelley) [inline pkg/protodep/protodep.proto]: I have a use case for this in a project which doesn't currently require solo-kit. It would be nice to have this in go-utils or even a standalone repo

File: pkg/code-generator/codegen/templates/kube/generate_script.go [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+
+${CODEGEN_PKG}/generate-groups.sh all \
+    ${CLIENT_PKG} \
+    ${APIS_PKG} \
+    {{ .ProjectConfig.Name }}:{{ .ProjectConfig.Version }} \
+    --output-base "${TEMP_DIR}"
+# Copy everything back.
+cp -a "${TEMP_DIR}/${ROOT_PKG}/." "${SCRIPT_ROOT}/.."
+
+`))
```
Comment (ilackarms) [inline pkg/code-generator/codegen/templates/kube/generate_script.go]: as mentioned, i think we should consider splitting out these new templates into a separate PR.  i'm not entirely sure we need it. i would prefer to run Codegen via libraries, as in autopilot: https://github.com/solo-io/autopilot/blob/master/codegen/util/deepcopy_gen.go#L55

File: pkg/code-generator/codegen/templates/kube/verify_codegen.go [hunk 30]
```diff
@@ -0,0 +1,47 @@
...
+cp -a "${TMP_DIFFROOT}"/* "${DIFFROOT}"
+if [[ $ret -eq 0 ]]
+then
+  echo "${DIFFROOT} up to date."
+else
+  echo "${DIFFROOT} is out of date. Please run 'make code-gen'"
+  exit 1
+fi
+
+`))
```
Comment (ilackarms) [inline pkg/code-generator/codegen/templates/kube/verify_codegen.go]: why do we need to generate this file with solo kit? 

i also don't see a `code-gen` make target anywhere.  where is this `verify_codegen.sh` coming from?

File: pkg/code-generator/collector/collector.go [hunk 31]
```diff
@@ -0,0 +1,284 @@
...
+	"github.com/solo-io/go-utils/log"
+	"github.com/solo-io/go-utils/stringutils"
+	"github.com/solo-io/solo-kit/pkg/code-generator/model"
+	"github.com/solo-io/solo-kit/pkg/code-generator/parser"
+	"github.com/solo-io/solo-kit/pkg/errors"
+	"golang.org/x/sync/errgroup"
+)
+
+type Collector interface {
+	CollectDescriptorsFromRoot(root string, skipDirs []string) ([]*model.DescriptorWithPath, error)
```
Comment (ilackarms) [inline pkg/code-generator/collector/collector.go]: please comment this interface. what does it do? how does it work?

File: pkg/code-generator/collector/collector.go [hunk 32]
```diff
@@ -0,0 +1,284 @@
...
+
+	cmd.Args = append(cmd.Args, "-o"+toFile, "--include_imports", "--include_source_info",
+		protoFile)
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return errors.Wrapf(err, "%v failed: %s", cmd.Args, out)
+	}
+	return nil
+}
```
Comment (ilackarms) [inline pkg/code-generator/collector/collector.go]: how was this file changed from the original implementation? or was code simply moved? it's hard to tell without being able to compare with the original code side by side

File: pkg/protodep/gomod.go [hunk 33]
```diff
@@ -0,0 +1,331 @@
...
+	LocalMatchers []string
+}
+
+// struct which represents a go module package in the module package list
+type Module struct {
+	ImportPath     string
+	SourcePath     string
+	Version        string
+	SourceVersion  string
+	Dir            string   // full path, $GOPATH/pkg/mod/
```
Comment (ilackarms) [inline pkg/protodep/gomod.go]: what is Dir? this PR would be a lot easier to review with comments for some of these new types / interfaces

File: pkg/protodep/protodep.go [hunk 34]
```diff
@@ -0,0 +1,46 @@
...
+	if err := opts.Validate(); err != nil {
+		return err
+	}
+	for _, v := range m.depFactories {
+		if err := v.Ensure(ctx, opts); err != nil {
+			return err
+		}
+	}
+	return nil
+}
```
Comment (ilackarms) [inline pkg/protodep/protodep.go]: since `protodep` package seems to be the core of this PR, i think it should be split into a separate PR and commented adequately so it can be reviewed properly. the scope of the PR in its current form is too large to evaluate the impact of all these changes at once

File: pkg/protodep/protodep.proto (hunk 28 above)
Comment (EItanya) [reply pkg/protodep/protodep.proto]: Planning on doing that once it's all in and working, as well as adding support for git repos

File: pkg/code-generator/codegen/templates/kube/doc_template.go (hunk 27 above)
Comment (EItanya) [reply pkg/code-generator/codegen/templates/kube/doc_template.go]: Yes we do because all of these files are generated

File: pkg/code-generator/collector/collector.go (hunk 31 above)
Comment (EItanya) [reply pkg/code-generator/collector/collector.go]: It does nothing differently from before, I made it an interface only so we could possibly add tests at a later date

File: pkg/code-generator/codegen/generator.go (hunk 26 above)
Comment (EItanya) [reply pkg/code-generator/codegen/generator.go]: We talked about this already. The old simple generation targets don't work because we can't reply on the GOPATH anymore. These scripts are based off of https://github.com/weaveworks/flagger/blob/master/hack/update-codegen.sh

File: generate.go (hunk 25 above)
Comment (EItanya) [reply generate.go]: I was thinking this as well, I can abstract away some of the basic calls. For instance if you look at the Gloo PR, it is much simpler. There will most likely have to be user additions of some kind no matter what though.

File: pkg/code-generator/codegen/generator.go [hunk 35]
```diff
@@ -188,6 +191,19 @@ func generateKubeFilesForProject(project *model.Project) (code_generator.Files,
 			Content:  content,
 		})
 	}
+	for suffix, tmpl := range map[string]*template.Template{
+		"hack/update-codegen.sh": kube.GenerateScriptTemplate,
```
[the remark this answers is not in the document]
Comment (EItanya) [reply pkg/code-generator/codegen/generator.go]: These are specifically for kubernetes code-generation

File: pkg/code-generator/docgen/funcs/template_funcs.go [hunk 36]
```diff
@@ -593,6 +593,9 @@ func (c *templateFunctions) forEachMessage(messagesToSkip map[string]bool) func(
 				continue
 			}
 
+			if msg.GetName() == "SInt32Rules" || msg.GetName() == "SInt64Rules" {
```
Comment (ilackarms) [inline pkg/code-generator/docgen/funcs/template_funcs.go]: remove
